include/ginkgo/core/config/data.hpp

```diff
@@ -0,0 +1,268 @@
+
+    template <tag_type>
+    struct tag {
+        using type = void;
+    };
+
+    /**
+     * Get the data with Type T
+     */
+    template <typename T>
+    inline T get() const
+    {
+        throw std::runtime_error("Not Supported");
+    }
+
+    template <tag_type T>
+    inline typename tag<T>::type get() const;
+
+    inline tag_type get_tag() const { return tag_; }
+
+private:
+    tag_type tag_;
+    std::string str_;
+    union {
+        long long int int_;
+        double double_;
+        bool bool_;
+    } u_;
+};
+
+template <>
+inline long long int data::get<long long int>() const
+{
+    assert(tag_ == data::tag_type::int_t);
+    return u_.int_;
+}
+
+template <>
+inline std::string data::get<std::string>() const
+{
+    assert(tag_ == data::tag_type::str_t);
+    return str_;
+}
+
+template <>
+inline double data::get<double>() const
+{
+    assert(tag_ == data::tag_type::double_t);
+    return u_.double_;
+}
+
+template <>
+inline bool data::get<bool>() const
+{
+    assert(tag_ == data::tag_type::bool_t);
+    return u_.bool_;
+}
+
+
+template <>
+struct data::tag<data::tag_type::bool_t> {
+    using type = bool;
+};
+
+template <>
+struct data::tag<data::tag_type::int_t> {
+    using type = long long int;
+};
+
+template <>
+struct data::tag<data::tag_type::double_t> {
+    using type = double;
+};
+
+template <>
+struct data::tag<data::tag_type::str_t> {
+    using type = std::string;
+};
+
+
+template <typename T>
+inline T get(const data& d)
+{
+    static_assert(std::is_same<T, std::string>::value ||
+                      std::is_same<T, long long int>::value ||
+                      std::is_same<T, double>::value ||
+                      std::is_same<T, bool>::value ||
+                      std::is_same<T, monostate>::value,
+                  "Not supported data type");
+    if (holds_alternative<T>(d)) {
+        return d.get<T>();
+    }
+    throw std::runtime_error(std::string("data does not holds the type ") +
+                             typeid(T).name());
+}
+
+
+template <typename T>
+inline bool holds_alternative(const data& d)
+{
+    static_assert(std::is_same<T, std::string>::value ||
+                      std::is_same<T, long long int>::value ||
+                      std::is_same<T, double>::value ||
+                      std::is_same<T, bool>::value ||
+                      std::is_same<T, monostate>::value,
+                  "Not supported data type");
+    if (std::is_same<T, std::string>::value) {
+        return d.get_tag() == data::tag_type::str_t;
+    } else if (std::is_same<T, long long int>::value) {
+        return d.get_tag() == data::tag_type::int_t;
+    } else if (std::is_same<T, double>::value) {
+        return d.get_tag() == data::tag_type::double_t;
+    } else if (std::is_same<T, bool>::value) {
+        return d.get_tag() == data::tag_type::bool_t;
+    } else if (std::is_same<T, monostate>::value) {
+        return d.get_tag() == data::tag_type::empty_t;
+    }
+}
```

Comment:
All of these can go into the source file

Reply:
no, moving into the source will lead template broken.

Reply:
they just need to be instantiated explicitly. Our header files are already way too full, we should try to keep implementations out of new files as much as possible.

Reply:
does it reduce the size? For example, holds_alternative takes 21 lines now, but the explicit instantiation (2*5) + default implementation (5) + the linebreak (5) also needs 20 lines

Reply:
Header files are supposed to contain declarations, not definitions. We are doing a really bad job at this (see #1105 for an attempt to improve things), but I would argue we can and should do better than this. Every time you include a file, all its functions have to be parsed and (often) be instantiated, which also leads to a lot of duplicate symbols (related to #715).

Reply:
Bumping this again, the headers should show the interface, not the implementation.